### DISCUSSION TITLE
check_drivesize and C drive exclusion // Unknown return

### DISCUSSION THREAD

**Discussion by brunssonM** [mickem/nscp#847] [Q&A]
Hello
I tried to have a check for all other drive than C drive, so i have add a exclude=C in my command, to have this one : 
/usr/lib64/nagios/plugins/check_centreon_nrpe3 -H HOSTNAME -p 5666 -t 60 -2 -P 8192  -c check_drivesize -a "drive=*" "perf-config=used(unit:GB)used %(ignored:true)" "filter=type = 'fixed' and name not regexp '.*System Reserved.*' and type not in('CDROM','REMOVABLE') and mounted=1" "warning=total_used>96%" "critical=total_used>96%" "exclude=C"

Under poller, if i run comand i have this return : 
OK: No drives found

But under web interface i have an UNKNOWN and not OK.
For server where there is an other drive than C it works (it return size of other drive), but for system with only C: drive, not an OK even if return is : OK: No drives found

Thanks a lot for your help, as i don't find solution and i don't want to create own check for this existing feature.

**Comment by mickem** [comment]
There are a lot of things going on here so if I missunderstood your problem please do let me know...
What I think the problem is is that when you try to check and find nothing the "default" return is unknown. As we expect a check to check something.


Does that make sense?

If not lets make an example:
And I am running this in the "test client" as that gives me the output I need to show, but it does not matter...
```
check_drivesize "filter=drive='C:\\'"
L        cli CRITICAL: CRITICAL C:\: 857.994GB/928.7GB used
L        cli  Performance data: 'C:\ used'=857.99433GB;742.95972;835.82968;0;928.69965 'C:\ used %'=92%;80;90;0;100
```
Here we can clearly seee that I need to clean my disk (or buy a new SSD).
Now the more important part here is that we get `CRITICAL: CRITICAL` twice.
Why you see the status Twice is that in the return text we print the status.
So the first `UNKNOWN` here is the actual result of the command and the second `UNKNOWN` is for historical reasons included in the text.

Now if I instead check a non existing drive called foobar:
```
check_drivesize "filter=drive='foobar'"
L        cli UNKNOWN: OK: No drives found
L        cli  Performance data:
```
What we get now is `UNKNOWN` AND `OK`.
This is super confusing and arguably a bug. But now the `UNKNOWN` is the actual result of the check and the `OK` is just part of the message.

The reason here is that we have two things happening:
1. The filter did not match anything so the filter thinks the result is ok and sets the text to "ok".
2. There were no results so we override the result from the filtering and return unknown.

To control this we can use the `empty-state` option:
```
check_drivesize "filter=drive='foobar'" empty-state=warning
L        cli WARNING: OK: No drives found
L        cli  Performance data:
```

Which might be more confusing, but in your case perhaps the "empty list" is ok?


That said the message should match the actual status here and this is a bug, but it is only a bug in the message so it is not afffecting your alerts.

**Comment by mickem** [comment]
(the bug mentioned above will be fixed in next version)
